### PR TITLE
Merge in from Gorkowski Dev

### LINKED ISSUE #473: Merge in from Gorkowski Dev
full merge in

### PR DESCRIPTION
**Pull request by Gorkowski**
Fixes #473

## Summary by Sourcery

Introduce new coagulation strategies and surface tension factories, enhance existing classes with improved methods and flexibility, fix bugs in mean free path calculations, and update documentation and tests to reflect these changes.

New Features:
- Introduce a new `CoagulationStrategy` module with various strategies for calculating coagulation rates, including `DiscreteSimple`, `DiscreteGeneral`, `ContinuousGeneralPDF`, and `ParticleResolved` strategies.
- Add a `SurfaceFactory` class to create surface tension strategy builders for calculating surface tension and the Kelvin effect for species in particulate phases.
- Implement a `VaporPressureFactory` class to create vapor pressure strategy builders for calculating vapor pressure of gas species.
- Add new functions for converting mass concentrations to mole and volume fractions in the `convert_mass_concentration` module.

Bug Fixes:
- Fix issues in the `mean_free_path` calculation by updating the method to use the correct dynamic viscosity and gas constant values.
- Correct the handling of temperature unit conversions in the `convert_units` function to ensure accurate conversions between Celsius, Fahrenheit, and Kelvin.

Enhancements:
- Refactor the `GasSpecies` class to support both single and array inputs for gas properties, improving flexibility and usability.
- Improve the `Stream` and `StreamAveraged` classes by adding methods for data manipulation, such as `__getitem__`, `__setitem__`, and `get_std`.
- Enhance the `CoulombEnhancement` class by adding detailed docstrings and improving the calculation methods for potential ratio and enhancement limits.
- Refactor the `vapor_pressure_builders` module to use a base `BuilderABC` class for better code organization and error handling.

Build:
- Update the `setup-python` action in GitHub workflows to cache pip dependencies, improving build efficiency.

CI:
- Add caching for pip dependencies in GitHub Actions workflows to speed up the continuous integration process.

Documentation:
- Update the documentation to include new notebooks on light scattering and equilibria, providing users with more examples and guides.
- Revise the `mean_free_path` notebook to reflect changes in the calculation method and improve clarity.

Tests:
- Add new tests for the `CoagulationStrategy` module to ensure the correctness of different coagulation strategies.
- Introduce tests for the `SurfaceFactory` class to verify the creation of surface tension strategies.
- Enhance existing tests for the `convert_mass_concentration` module to cover new functionalities and edge cases.

Chores:
- Update the version number to `0.0.17.dev0` to reflect ongoing development changes.
- Clean up and organize imports across multiple modules for better readability and maintainability.

### CODE REVIEW

File: particula/next/gas/vapor_pressure_builders.py [hunk 1]
```diff
@@ -13,7 +14,7 @@
 logger = logging.getLogger("particula")
 
 
-class AntoineBuilder():
+class AntoineBuilder(BuilderABC):
```
[the remark this answers is not in the document]
Comment (Gorkowski) [reply particula/next/gas/vapor_pressure_builders.py:17]: Create issue

File: particula/data/process/ml_analysis/generate_and_train_2mode_sizer.py [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+    # Evaluate the pipeline with noisy data
+    print("Evaluating pipeline with noisy data:")
+    evaluate_pipeline(pipeline_noisy, x_test_noisy, y_test_noisy)
+
+    # Save the pipeline with noisy data
+    save_pipeline(pipeline, save_path)
+
+
+# pylint: disable=too-many-locals
+def lognormal_2mode_ml_guess(
```
[the remark this answers is not in the document]
Comment (Gorkowski) [reply particula/data/process/ml_analysis/generate_and_train_2mode_sizer.py:455]: Generate a GitHub issue from a review comment

File: particula/data/process/ml_analysis/generate_and_train_2mode_sizer.py [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+    Arguments:
+        filename: The filename to load the pipeline from.
+
+    Returns:
+        The loaded pipeline.
+    """
+    return joblib.load(filename)  # type: ignore
+
+
+def load_and_cache_pipeline(filename: str) -> Pipeline:
```
[the remark this answers is not in the document]
Comment (Gorkowski) [reply particula/data/process/ml_analysis/generate_and_train_2mode_sizer.py:368]: Generate a GitHub issue from a review comment:

**Comment by Gorkowski** [comment]
I'm just wanted to create some issues for most of the soucery comments so we don't forget

I'm good with it now
